raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,162 @@
+import json
+
+from web3 import Web3
+
+from raiden.exceptions import InvalidDBData
+from raiden.storage.sqlite import EventRecord, SQLiteStorage, StateChangeRecord
+from raiden.utils.typing import Any, BlockNumber, Dict
+
+SOURCE_VERSION = 18
+TARGET_VERSION = 19
+
+
+class BlockHashCache(object):
+    """A small cache for blocknumber to blockhashes to optimize this migration a bit
+
+    This cache lives only during the v18->v19 migration where numerous RPC calls are
+    expected to be made, many of them probably querying the same block_number -> block_hash.
+    To save RPC calls and make it a bit faster we keep this short-lived cache for the duration
+    of the migration
+    """
+    def __init__(self, web3: Web3):
+        self.web3 = web3
+        self.mapping = {}
+
+    def get(self, block_number: BlockNumber) -> str:
+        """Given a block number returns the hex representation of the blockhash"""
+        if block_number in self.mapping:
+            return self.mapping[block_number]
+
+        block_hash = self.web3.eth.getBlock(block_number)['hash']
+        block_hash = block_hash.hex()
+        self.mapping[block_number] = block_hash
+        return block_hash
+
+
+def _add_blockhash_to_state_changes(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to ContractReceiveXXX and ActionInitChain state changes"""
+
+    for state_changes_batch in storage.batch_query_state_changes(batch_size=500):
+
+        updated_state_changes = []
+        for state_change in state_changes_batch:
+            data = json.loads(state_change.data)
+            affected_state_change = (
+                'raiden.transfer.state_change.ContractReceive' in data['_type'] or
+                'raiden.transfer.state_change.ActionInitChain' in data['_type']
+            )
+            if not affected_state_change:
+                continue
+
+            assert 'block_hash' not in data, 'v18 state changes cant contain blockhash'
+            block_number = int(data['block_number'])
+            data['block_hash'] = cache.get(block_number)
```

Comment:
This is potentially really slow, because we are doing one rpc call at the time, IOW this is a blocking call, having a pool of `greenlet.Pool` and using `imap` will preserve the order, limit the number of concurrent RPC calls and allow the whole thing to update in parallel.

Synchronization at the cache level is not really necessary, since we know that the interpreter state will be consistent, we may just do a few useless calls here and there depending on the way the greenlets are scheduled. (Well, this is true up to the confirmed block, unconfirmed blocks are a problem of their own, but at the time we only have confirmed blocks in the database)

Reply:
Hey @hackaugusto ... I totally see how this will be slow but did not see any immediate way of making it faster and the PR is growing bigger.

I made an attempt with `gevent.Pool` and `imap` here:
https://github.com/raiden-network/raiden/pull/3550/commits/25ebfd1c8da235563e9c92a4321e584d2ed50e72

Is this close to what you were thinking? Please review carefully as I am not experienced with gevent.Pool and the documentation and examples are seriously lacking.

Another alternative and perhaps the more correct approach would be to create a batch of block number queries and see if web3 could batch them all in one single RPC call which will return many results. We know that normal web3 can. But `web3.py` seems to still have an open issue about this [here](https://github.com/ethereum/web3.py/issues/832)